Use non-blocking copy for creation of lazy tensors in TS backend impl

wconstab opened the pull request:
- blocking here makes cpu tracing thread wait for the transfer and cuda
  sync, before continuing on to trace more stuff.  we want to overlap
  cpu tracing with doing the copy in the background
- blocking is not required if the tensor is already on the cuda device,
  but it is if the tensor is on the cpu device since the cpu thread
  could modify the tensor while it is being copied asynchronously.
- we make an exception for numel()=1 tensors since (presumably?) they
  get copied as a stack variable safely, but maybe this is not true.

wconstab committed:
Use non-blocking copy for creation of lazy tensors in TS backend impl

- blocking here makes cpu tracing thread wait for the transfer and cuda
  sync, before continuing on to trace more stuff.  we want to overlap
  cpu tracing with doing the copy in the background

- blocking is not required if the tensor is already on the cuda device,
  but it is if the tensor is on the cpu device since the cpu thread
  could modify the tensor while it is being copied asynchronously.
- we make an exception for numel()=1 tensors since (presumably?) they
  get copied as a stack variable safely, but maybe this is not true.

wconstab commented:
A couple questions (cc @bdhirsh @ngimel)
1) why is `is_wrapped_number()` false for the `math.sqrt(bias_correction2)` term here: https://www.internalfb.com/code/fbsource/[ffdbc4cb2befea2584e7698a58d0fc72712326b1]/fbcode/caffe2/torch/optim/_functional.py?lines=104

2) whether it is safe to do an async copy to cuda for these numel()==1 cases (or, for actual is_wrapped_number() cases)

bdhirsh commented:
> 1. why is `is_wrapped_number()` false for the `math.sqrt(bias_correction2)` term here: [https://www.internalfb.com/code/fbsource/[ffdbc4cb2befea2584e7698a58d0fc72712326b1]/fbcode/caffe2/torch/optim/_functional.py?lines=104](https://www.internalfb.com/code/fbsource/%5Bffdbc4cb2befea2584e7698a58d0fc72712326b1%5D/fbcode/caffe2/torch/optim/_functional.py?lines=104)


In the code linked above, `bias_correction2` is computed as `1 - beta2 ** step` (beta2 is a `float`, step is an `int`). Seems like there's no relation to torch tensors here?

> 2. whether it is safe to do an async copy to cuda for these numel()==1 cases (or, for actual is_wrapped_number() cases)

No idea about this, Natalia can probably answer that better than I can.

wconstab commented:
> In the code linked above, `bias_correction2` is computed as `1 - beta2 ** step` (beta2 is a `float`, step is an `int`). Seems like there's no relation to torch tensors here?

Oh, yea, `math.sqrt(bias_correction2)` is definitely a python float.  But then in the course of performing 
`exp_avg_sq.sqrt() / math.sqrt(bias_correction2)`, the div operator seems to be wrapping it in a tensor, but that tensor claims to not be a wrapped number, hence my question.

wconstab committed:
Make numel()==1 path safely async

- doing a .to (nonblocking) cpu to cuda is potentially dangerous
  even for single-elem tensors; but fill_ on cuda tensors is an
  async operation, and .item() on cpu singletons tensors is fast

Krovatkin commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/backend_impl.cpp:
```diff
@@ -69,7 +69,16 @@ class TSBackendImpl : public torch::lazy::BackendImplInterface {
       const torch::lazy::BackendDevice& device) const override {
     at::TensorOptions options = tensor.options().device(default_device_type_.c10Type());
-    return std::make_shared<TSData>(tensor.to(options), shape, device);
+    if (tensor.device().type() == default_device_type_.c10Type()) {
```
wouldn't this be unsafe on CPU to CPU, or will it be sync on CPU regardless of `non_blocking` ?

wconstab replied:
yea that's a good question. i assumed cpu is always blocking but i need to confirm that.

i guess i could just make it a special case for cuda and be sure.

Krovatkin replied:
yes please!!! Could we please stick in  `&& tensor.device().type() == c10::kCUDA` in the first conditional for now until we confirm it works for CPU as well.

Krovatkin reviewed (approved):
:shipit:

wconstab committed:
Only use non-blocking transfer for cuda

Krovatkin commented on lazy_tensor_core/lazy_tensor_core/csrc/ts_backend/backend_impl.cpp:
```diff
@@ -69,7 +69,8 @@ class TSBackendImpl : public torch::lazy::BackendImplInterface {
     at::TensorOptions options = tensor.options().device(default_device_type_.c10Type());
-    if (tensor.device().type() == default_device_type_.c10Type()) {
+    if (tensor.device().type() == default_device_type_.c10Type() &&
+        default_device_type_.c10Type() == at::kCUDA) {
```
❤️

wconstab committed:
Fix sizes issue ([] vs [1])
